gprofiler/py_package_version.py

```diff
@@ -0,0 +1,190 @@
+    metadata = _get_metadata(dist)
+    if metadata:
+        # The metadata should NEVER be missing the Name: key, but if it somehow
+        # does, fall back to the known canonical name.
+        return metadata.get("Name", dist.project_name)
+    return None
+
+
+_warned_no__normalized_cached = False
+
+
+def get_versions(modules_paths: List[str], pid: int):
+    """Return a dict with module_path: (package_name, version). If couldn't
+    determine the version the value is None.
+
+    modules_paths must be absoulte. pid is required to access the path via
+    /proc/{pid}/root/.
+
+    Given a path to a module, it's not trivial to determine its package as some
+    packages contain modules with completely unrelated names (e.g. the module
+    pkg_resources used in this function is a part of the setuptools package).
+    This function gathers info for all the packages in the (dist|site)-packages
+    directory in each path, including a list of all the files associated with
+    each package. This list is searched for the given module path.
+    """
+    result = dict.fromkeys(modules_paths)
+
+    # A little monkey patch to prevent pkg_resources from converting "/proc/{pid}/root/" to "/".
+    # This function resolves symlinks and makes paths absolute for comparison purposes which isn't required
+    # for our usage.
+    if hasattr(pkg_resources, "_normalize_cached"):
+        original__normalize_cache = pkg_resources._normalize_cached
+        pkg_resources._normalize_cached = lambda path: path
+    else:
+        global _warned_no__normalized_cached
+        if not _warned_no__normalized_cached:
+            # Log only once so we don't spam the log
+            logger.warning("Cannot get modules version, pkg_resources has no '_normalize_cached' attribute")
+            _warned_no__normalized_cached = True
+
+        # Not much that we can do, pkg_resources.find_distributions won't work properly
+        return result
+
+    path_to_dist = {}
+
+    for path in modules_paths:
+        if not path.startswith("/"):
+            continue
+
+        if path not in path_to_dist:
+            packages_path = _get_packages_dir(path)
+            if packages_path is None:
+                # This module is (probably) not part of a package
+                continue
+            packages_path = _convert_to_proc_root_path(packages_path, pid)
+
+            # Make sure to catch any exception. If something goes wrong just don't get the version, we shouldn't
+            # interfere with gProfiler
+            try:
+                for dist in pkg_resources.find_distributions(packages_path):
+                    files_iter = _files_from_record(dist) or _files_from_legacy(dist)
+                    if files_iter is not None:
+                        path_to_dist.update(
+                            dict.fromkeys((os.path.join(packages_path, file) for file in files_iter), dist)
+                        )
+            except Exception:
+                pass
```

Comment:
I think we should log here, but I'm afraid the log will be spammed heavily in case the same error occur for every module.
@Jongy @guybortnikov WDYT?

Reply:
I do think we should log here, but we should catch more specific exceptions + don't try more than 1 time on a failed unit (failed unit may be for example a "dist", so if we already failed on a certain "dist" we shouldn't try again)

Reply:
Any specific exceptions are caught before this `try`, so what we're interested in here are actually the exceptions that aren't expected.
With this implementation units are different between processes, so the log we'll be spammed nonetheless. I suggest we add a log here once we implement the cache.